File: surface/widgets/compose/components/widgets/concepts/llm-conversation/ArtifactPanel.kt
package com.clef.surface.widgets.concepts.llmconversation

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalClipboardManager
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.delay

// --- State machine ---

enum class ArtifactPanelState { Open, Copied, Fullscreen, Closed }

sealed class ArtifactPanelEvent {
    object Copy : ArtifactPanelEvent()
    object Fullscreen : ArtifactPanelEvent()
    object Close : ArtifactPanelEvent()
    object VersionChange : ArtifactPanelEvent()
    object CopyTimeout : ArtifactPanelEvent()
    object ExitFullscreen : ArtifactPanelEvent()
    object Reopen : ArtifactPanelEvent()
}

fun artifactPanelReduce(state: ArtifactPanelState, event: ArtifactPanelEvent): ArtifactPanelState = when (state) {
    ArtifactPanelState.Open -> when (event) {
        is ArtifactPanelEvent.Copy -> ArtifactPanelState.Copied
        is ArtifactPanelEvent.Fullscreen -> ArtifactPanelState.Fullscreen
        is ArtifactPanelEvent.Close -> ArtifactPanelState.Closed
        is ArtifactPanelEvent.VersionChange -> ArtifactPanelState.Open
        else -> state
    }
    ArtifactPanelState.Copied -> when (event) {
        is ArtifactPanelEvent.CopyTimeout -> ArtifactPanelState.Open
        else -> state
    }
    ArtifactPanelState.Fullscreen -> when (event) {
        is ArtifactPanelEvent.ExitFullscreen -> ArtifactPanelState.Open
        is ArtifactPanelEvent.Close -> ArtifactPanelState.Closed
        else -> state
    }
    ArtifactPanelState.Closed -> when (event) {
        is ArtifactPanelEvent.Reopen -> ArtifactPanelState.Open
        else -> state
    }
}

// --- Types ---

enum class ArtifactType(val icon: String, val label: String) {
    Code("\uD83D\uDCBB", "Code"),
    Document("\uD83D\uDCC4", "Document"),
    Image("\uD83D\uDDBC", "Image"),
    Html("\uD83C\uDF10", "HTML")
}

// --- Component ---

@Composable
fun ArtifactPanel(
    content: String,
    artifactType: ArtifactType,
    title: String,
    modifier: Modifier = Modifier,
    language: String? = null,
    showVersions: Boolean = true,
    currentVersion: Int = 1,
    totalVersions: Int = 1,
    onVersionChange: (Int) -> Unit = {},
    onClose: () -> Unit = {},
    onCopy: () -> Unit = {},
    onDownload: () -> Unit = {}
) {
    var state by remember { mutableStateOf(ArtifactPanelState.Open) }
    val clipboardManager = LocalClipboardManager.current

    // Copy timeout
    LaunchedEffect(state) {
        if (state == ArtifactPanelState.Copied) {
            delay(2000)
            state = artifactPanelReduce(state, ArtifactPanelEvent.CopyTimeout)
        }
    }

    // Notify parent on close
    LaunchedEffect(state) {
        if (state == ArtifactPanelState.Closed) {
            onClose()
        }
    }

    if (state == ArtifactPanelState.Closed) return

    val showVersionBar = showVersions && totalVersions > 1

    Column(
        modifier = modifier
            .fillMaxWidth()
            .semantics { contentDescription = "Artifact panel: $title" }
    ) {
        // Header
        Row(
            verticalAlignment = Alignment.CenterVertically,
            modifier = Modifier
                .fillMaxWidth()
                .padding(8.dp)
        ) {
            // Type badge
            Text("${artifactType.icon} ${artifactType.label}", fontSize = 12.sp, modifier = Modifier.padding(end = 8.dp))

            // Title
            Text(title, fontWeight = FontWeight.SemiBold, fontSize = 14.sp, modifier = Modifier.weight(1f))

            // Toolbar
            Row(horizontalArrangement = Arrangement.spacedBy(4.dp)) {
                // Copy
                TextButton(onClick = {
                    clipboardManager.setText(AnnotatedString(content))
                    state = artifactPanelReduce(state, ArtifactPanelEvent.Copy)
                    onCopy()
                }) {
                    Text(if (state == ArtifactPanelState.Copied) "Copied!" else "Copy", fontSize = 12.sp)
                }

                // Download
                TextButton(onClick = { onDownload() }) {
                    Text("Download", fontSize = 12.sp)
                }

                // Fullscreen toggle
                TextButton(onClick = {
                    state = if (state == ArtifactPanelState.Fullscreen)
                        artifactPanelReduce(state, ArtifactPanelEvent.ExitFullscreen)
                    else
                        artifactPanelReduce(state, ArtifactPanelEvent.Fullscreen)
                }) {
                    Text(
                        if (state == ArtifactPanelState.Fullscreen) "Exit Fullscreen" else "Fullscreen",
                        fontSize = 12.sp
                    )
                }

                // Close
                TextButton(onClick = {
                    state = artifactPanelReduce(state, ArtifactPanelEvent.Close)
                }) {
                    Text("Close", fontSize = 12.sp)
                }
            }
        }

        // Version bar
        if (showVersionBar) {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(8.dp),
                modifier = Modifier.padding(horizontal = 12.dp, vertical = 4.dp)
            ) {
                TextButton(
                    onClick = {
                        if (currentVersion > 1) {
                            state = artifactPanelReduce(state, ArtifactPanelEvent.VersionChange)
                            onVersionChange(currentVersion - 1)
                        }
                    },
                    enabled = currentVersion > 1
                ) { Text("\u2039", fontSize = 14.sp) }

                Text("Version $currentVersion of $totalVersions", fontSize = 12.sp)

                TextButton(
                    onClick = {
                        if (currentVersion < totalVersions) {
                            state = artifactPanelReduce(state, ArtifactPanelEvent.VersionChange)
                            onVersionChange(currentVersion + 1)
                        }
                    },
                    enabled = currentVersion < totalVersions
                ) { Text("\u203A", fontSize = 14.sp) }
            }
        }

        HorizontalDivider()

        // Content area
        when (artifactType) {
            ArtifactType.Code -> {
                Column(Modifier.padding(12.dp)) {
                    language?.let {
                        Text(it.uppercase(), fontSize = 11.sp, color = MaterialTheme.colorScheme.onSurfaceVariant, modifier = Modifier.padding(bottom = 8.dp))
                    }
                    Text(content, fontFamily = FontFamily.Monospace, fontSize = 13.sp)
                }
            }
            ArtifactType.Document -> {
                Text(content, modifier = Modifier.padding(16.dp), fontSize = 14.sp, lineHeight = 22.sp)
            }
            ArtifactType.Image -> {
                Box(
                    contentAlignment = Alignment.Center,
                    modifier = Modifier.fillMaxWidth().padding(16.dp)
                ) {
                    Text("[Image: $title]", color = MaterialTheme.colorScheme.onSurfaceVariant)
                }
            }
            ArtifactType.Html -> {
                Column(Modifier.padding(16.dp)) {
                    Text("HTML preview is sandboxed. Raw HTML is not rendered directly.",
                        fontSize = 13.sp, color = MaterialTheme.colorScheme.onSurfaceVariant,
                        modifier = Modifier.padding(bottom = 8.dp))
                    Text(content, fontFamily = FontFamily.Monospace, fontSize = 13.sp)
                }
            }
        }
    }
}

File: surface/widgets/swiftui/components/widgets/concepts/llm-conversation/ArtifactPanel.swift
import SwiftUI

// MARK: - State Machine

enum ArtifactPanelState { case open, copied, fullscreen, closed }
enum ArtifactPanelEvent { case copy, fullscreen, close, versionChange, copyTimeout, exitFullscreen, openPanel }

func artifactPanelReduce(state: ArtifactPanelState, event: ArtifactPanelEvent) -> ArtifactPanelState {
    switch state {
    case .open:
        switch event {
        case .copy: return .copied
        case .fullscreen: return .fullscreen
        case .close: return .closed
        case .versionChange: return .open
        default: return state
        }
    case .copied:
        if case .copyTimeout = event { return .open }
        return state
    case .fullscreen:
        switch event {
        case .exitFullscreen: return .open
        case .close: return .closed
        default: return state
        }
    case .closed:
        if case .openPanel = event { return .open }
        return state
    }
}

// MARK: - Types

enum ArtifactType: String { case code, document, image, html }

private let typeIcons: [ArtifactType: String] = [.code: "\u{1F4BB}", .document: "\u{1F4C4}", .image: "\u{1F5BC}", .html: "\u{1F310}"]
private let typeLabels: [ArtifactType: String] = [.code: "Code", .document: "Document", .image: "Image", .html: "HTML"]

// MARK: - View

struct ArtifactPanelView: View {
    let content: String
    let artifactType: ArtifactType
    let title: String
    var language: String?
    var showVersions: Bool = true
    var currentVersion: Int = 1
    var totalVersions: Int = 1
    var onVersionChange: ((Int) -> Void)?
    var onClose: (() -> Void)?
    var onCopy: (() -> Void)?

    @State private var widgetState: ArtifactPanelState = .open
    @State private var copyTimer: Timer?

    private var showVersionBar: Bool { showVersions && totalVersions > 1 }

    var body: some View {
        if widgetState == .closed { EmptyView() }
        else {
            VStack(alignment: .leading, spacing: 0) {
                // Header
                HStack(spacing: 8) {
                    Text(typeIcons[artifactType] ?? "").font(.system(size: 12))
                    Text(typeLabels[artifactType] ?? "").font(.system(size: 12))
                    Text(title).fontWeight(.semibold).lineLimit(1)
                    Spacer()

                    Button(widgetState == .copied ? "Copied!" : "Copy") { handleCopy() }
                        .font(.caption)
                    Button(widgetState == .fullscreen ? "Exit Fullscreen" : "Fullscreen") {
                        widgetState = widgetState == .fullscreen
                            ? artifactPanelReduce(state: widgetState, event: .exitFullscreen)
                            : artifactPanelReduce(state: widgetState, event: .fullscreen)
                    }.font(.caption)
                    Button("Close") {
                        widgetState = artifactPanelReduce(state: widgetState, event: .close)
                        onClose?()
                    }.font(.caption)
                }
                .padding(.horizontal, 12).padding(.vertical, 8)

                // Version bar
                if showVersionBar {
                    HStack(spacing: 8) {
                        Button("\u{2039}") { if currentVersion > 1 { onVersionChange?(currentVersion - 1) } }
                            .disabled(currentVersion <= 1)
                        Text("Version \(currentVersion) of \(totalVersions)").font(.system(size: 12))
                        Button("\u{203A}") { if currentVersion < totalVersions { onVersionChange?(currentVersion + 1) } }
                            .disabled(currentVersion >= totalVersions)
                    }
                    .padding(.horizontal, 12).padding(.vertical, 4)
                }

                Divider()

                // Content
                ScrollView {
                    switch artifactType {
                    case .code:
                        VStack(alignment: .leading, spacing: 4) {
                            if let lang = language {
                                Text(lang.uppercased()).font(.system(size: 11)).foregroundColor(.secondary)
                            }
                            Text(content).font(.system(size: 13, design: .monospaced)).textSelection(.enabled)
                        }.padding(12)
                    case .document:
                        Text(content).font(.system(size: 14)).padding(16).textSelection(.enabled)
                    case .image:
                        // In a headless component, show the URL/path
                        Text(content).font(.system(size: 13)).foregroundColor(.secondary).padding(16)
                    case .html:
                        VStack(alignment: .leading, spacing: 8) {
                            Text("HTML preview is sandboxed.").font(.caption).foregroundColor(.secondary).italic()
                            Text(content).font(.system(size: 13, design: .monospaced)).textSelection(.enabled)
                        }.padding(12)
                    }
                }
            }
            .accessibilityElement(children: .contain)
            .accessibilityLabel("Artifact: \(title)")
        }
    }

    private func handleCopy() {
        #if os(macOS)
        NSPasteboard.general.clearContents()
        NSPasteboard.general.setString(content, forType: .string)
        #endif
        widgetState = artifactPanelReduce(state: widgetState, event: .copy)
        onCopy?()
        copyTimer?.invalidate()
        copyTimer = Timer.scheduledTimer(withTimeInterval: 2, repeats: false) { _ in
            widgetState = artifactPanelReduce(state: widgetState, event: .copyTimeout)
        }
    }
}
